skimage/filters/_sparse.py

```diff
@@ -0,0 +1,99 @@
+import numpy as np
+from ._sparse_cy import _correlate_sparse_offsets
+from collections.abc import Iterable
+
+
+def _validate_window_size(axis_sizes):
+    """Ensure all sizes in ``axis_sizes`` are odd.
+
+    Parameters
+    ----------
+    axis_sizes : iterable of int
+
+    Raises
+    ------
+    ValueError
+        If any given axis size is even.
+    """
+    for axis_size in axis_sizes:
+        if axis_size % 2 == 0:
+            msg = ('Window size for `threshold_sauvola` or '
+                   '`threshold_niblack` must not be even on any dimension. '
+                   'Got {}'.format(axis_sizes))
+            raise ValueError(msg)
+
+
+def _to_np_mode(mode):
+    """Convert padding modes from `ndi.correlate` to `np.pad`."""
+    mode_translation_dict = dict(nearest='edge', reflect='symmetric',
+                                 mirror='reflect')
+    if mode in mode_translation_dict:
+        mode = mode_translation_dict[mode]
+    return mode
+
+
+def correlate_sparse(image, kernel, mode='reflect'):
+    """Compute valid cross-correlation of `padded_array` and `kernel`.
+
+    This function is *fast* when `kernel` is large with many zeros.
+
+    See ``scipy.ndimage.correlate`` for a description of cross-correlation.
+
+    Parameters
+    ----------
+    image : ndarray, dtype float, shape (M, N,[ ...,] P)
+        The input array. If mode is 'valid', this array should already be
+        padded, as a margin of the same shape as kernel will be stripped
+        off.
+    kernel : ndarray, dtype float shape (Q, R,[ ...,] S)
+        The kernel to be correlated. Must have the same number of
+        dimensions as `padded_array`. For high performance, it should
+        be sparse (few nonzero entries).
+    mode : string, optional
+        See `scipy.ndimage.correlate` for valid modes. 
+        Additionally, mode 'valid' is accepted, in which case no padding is 
+        applied and the result is the result for the smaller image for which 
+        the kernel is entirely inside the original data.
+
+    Returns
+    -------
+    result : array of float, shape (M, N,[ ...,] P)
+        The result of cross-correlating `image` with `kernel`. If mode
+        'valid' is used, the resulting shape is (M-Q+1, N-R+1,[ ...,] P-S+1).
+    """
+    if not isinstance(kernel, np.ndarray):
+        msg = '`correlate_sparse` kernel must be an numpy array object.'
+        raise ValueError(msg)
```

Comment:
Can't we accept array like inputs for `kernel`?
```suggestion
    kernel = np.asarray(kernel)
```

Reply:
@rfezzani we had this discussion before and based on #4407  we decided to go for ndarray. https://github.com/scikit-image/scikit-image/pull/4912#discussion_r475174333 @jni please correct me if I am wrong.

Reply:
We can leave this decision for later. Honestly the entire ecosystem is having massive headaches with "array like" when it comes to typing, so I think I'd rather move to the future where everyone is responsible for providing actual NumPy arrays.

Again, we can revisit this decision later, I don't think it should hold up the PR.

Reply:
> Honestly the entire ecosystem is having massive headaches with "array like" 

Let's assume then that `kernel` is an array and avoid the check. I don't see such validation else where in `skimage`...

Reply:
I just merged this suggestion. The problem with assuming is that the code will fail in a rather obscure way (list object has no attribute shape).